### PR TITLE
Bootstrap the sections exploded view

### PR DESCRIPTION
**Pull request by youknowriad**
Related #39281 #40319

## What?

We want to explore an "exploded block view" where the blocks (and sections) of the top level are separated and the actions for each block/sections are separate from the regular toolbar.

See https://github.com/WordPress/gutenberg/issues/39281#issuecomment-1084648592

This PR just bootstraps that mode to allow us to iterate on it.

## Why?

I think the reasoning is to make it easy to build templates from scratch by leveraging ready to use patterns.

## Notes and questions?

There are already some questions that we need to answer in this initial PR:

 - Do we want this view to be available as a reusable "view" for each block editor. Meaning should we implement it as part of the "block-editor" package? This would mean adding a new component to be exported from "block-editor" package. Alternatively we start with a specific component in `edit-site` leveraging existing components (BlockPreview...). I've started with the latter for now but I'd love to hear thoughts.
 - The proposed design leverages an animation between the two modes. At this point, I'm not really certain that this would ever be possible to be honest. The main reason is that each "box" is going to be a separate iframe because it needs the editor styles to apply inside the box but not bleed outside, while in the default view, the whole canvas is a single iframe. Meaning if we were to implement an animation we need to animate between two completely different DOM elements rendered in different iframes. Right now `layoutId` based animation from framer-motion (which are our most promising solution) don't seem to work inside [a single iframe](https://href.li/?https://codesandbox.io/s/sandpack-project-forked-e83z31?file=/App.js), so cross-iframe is even less likely to be possible.
- The exploded view means all blocks render sequentially, meaning if we have “aligned” components at the top level, they’ll probably lose their alignment styles.
- I wonder if we can use the same design (exploded view) for the current “select mode” to avoid a proliferation of modes. It seems their goals are quiet similar. (manipulate blocks as units). I'm starting separate but there's an opportunity to unify later.
- Editing of blocks is disabled in exploded view (they are previews).
- Sometimes when switching back and forth between exploded and regular views, the iframe height kind of resets to 0 for some of the root blocks. I think that's mostly like a bug in `BlockPreview` component that we need to find and fix.
- One important caviat for this view as well as that it's kind of only suited for themes with a certain design: Themes with a vertical list of blocks at the root level. For instance, if you try it in a theme like "Kubrick 2", it doesn't work, the whole site is inside "one root" block. I don't know how concerning is that.

## Testing Instructions

1- Open the site editor
2- Click the "exploded view" button on the header.

### CODE REVIEW

**Comment by jasmussen** [comment]
Even in this early state, the exploded view already offers a clear delineation between top level sections, big and convenient points for inserting sections, a distinct context for suggesting particular patterns (headers, sections, footers), a better overview of large page-sized patterns, and a mode that intrinsically could offer intuitive drag and drop. I think there's something to this.

> Do we want this view to be available as a reusable "view" for each block editor. 

Depends! I imagine it would be useful at the very least for page editing, especially in helping choose page templates, or page-size patterns. It seems prudent to start in the site editor.

> The proposed design leverages an animation between the two modes. At this point, I'm not really certain that this would ever be possible to be honest. 

I'm curious what others feel, but I'd think the animation to a crucial part of this. Are there any creative solutions we can think of? Instead of animating the iframe itself, could we animate just the contents? I understand there would likely be some challenges with background colors, collapsing margins and such, but those seem worth delving into on their own.

> The exploded view means all blocks render sequentially, meaning if we have “aligned” components at the top level, they’ll probably lose their alignment styles.

You mean floats here? I'd think that acceptable.

> I wonder if we can use the same design (exploded view) for the current “select mode”

Agreed, there's enough overlap that it could potentially be the same, though I'd push any experiments to merge the two a bit in front of us still.

> Editing of blocks is disabled in exploded view (they are previews).

👌 — this would allow us to make the entire thing into a drag/drop'able unit, right?

> One important caviat for this view as well as that it's kind of only suited for themes with a certain design: Themes with a vertical list of blocks at the root level. For instance, if you try it in a theme like "Kubrick 2", it doesn't work, the whole site is inside "one root" block. I don't know how concerning is that.

At the moment there is a "focus" feature that lets you edit a template part in isolation. You click the template part, click "Edit" from the ellipsis menu, and you're editing that part in isolation, with dark clearance around it.

To an extent, I'm thinking of this interface as the same as that — a "focused" view. Technical challenges aside (I'm sure they are legion), we could make this view contextual; select a group block and click the button, and it explodes just the blocks inside in an isolated view, just like isolated template part editing. These are vague ideas, to be clear, but it's all to say that if we can get this to a place that makes us happy for vertical themes, there are options we can explore other layouts as well.

**Comment by jasmussen** [comment]
Exploring some icon options as an alternative to using the Stack icon, I don't really love any of these yet: 

a couple of doodles

**Comment by jameskoster** [comment]
> One important caviat for this view as well as that it's kind of only suited for themes with a certain design: Themes with a vertical list of blocks at the root level. For instance, if you try it in a theme like "Kubrick 2", it doesn't work, the whole site is inside "one root" block. I don't know how concerning is that.

Perhaps there are some rules we can apply that determine whether or not this mode is enabled? One could be; 

If there are ≤1 root blocks in the current view, disable exploded view. 

That way – as Joen says – if you entered focus mode for the single root block, exploded view would become enabled.

The challenge then is how to handle _horizontal_ layouts, I think you mentioned to this in the OP. For example consider this structure:

Screenshot 2022-04-13 at 17 35 58

If I enter focus mode for the root Row, then entered exploded view, it wouldn't work. So we'd probably also need a rule that says; 

If the container orientation is horizontal, then disable exploded view.

**Comment by youknowriad** [comment]
I've been thinking about this PR, it's clear that we won't be able to implement the mode entirely in a single PR and since it's a new button and don't have any backward compatibility impact or whatsoever, I was thinking what would be the minimal accepted version for this to land to allow us to iterate on it quickly.

I know the mode needs:

 - Click to select a block
 - A dedicate toolbar with just the movers and the name of the block
 - Pattern switchers for "semantic sections" on the side
 - Animation between the modes
 - Drag and drop to move top level blocks
 - Double click to go back to "normal" mode?
 - Find the right icon in the toolbar
 - A different design for the list view (focused on sections/ root level blocks)
 - A different design for the inserter (focused on patterns)
 - Remove button on the side

Am I missing anything here?

Now, from that list I think the MVP could include the following and leave the rest for dedicated PRs.

 - Click to select a block
 - A dedicate toolbar with just the movers and the name of the block
 - Double click to go back to "normal" mode?
 - Find the right icon in the toolbar

What do you think?

**Comment by jameskoster** [comment]
Would it be possible to also include some basic logic around when the mode is enabled or not? 

We should think about how to handle panels like Inserter / List view / Inspector in the mean time as well, the current disabled state feels a bit strange to me. It might be simpler to just force-close those panels when you enter exploded mode. Then if you click one in exploded mode, you just return to "normal" and the panel opens.

> Click to select a block
> Double click to go back to "normal" mode?

It would be neat to try this as well, but the original idea was that the toolbar(s) would appear on hover, and a single click on any section would return to "normal" mode. This is likely one of those things we'll need to "feel" to make a decision on.

**Comment by youknowriad** [comment]
> Would it be possible to also include some basic logic around when the mode is enabled or not?

Yes, I mean to include that. I think for now since it's only enabled in site editor root which doesn't have any "layout", we can just check the number of blocks.

> We should think about how to handle panels like Inserter / List view / Inspector in the mean time as well, the current disabled state feels a bit strange to me. It might be simpler to just force-close those panels when you enter exploded mode. Then if you click one in exploded mode, you just return to "normal" and the panel opens.

We can try things and see how it feels.

> It would be neat to try this as well, but the original idea was that the toolbar(s) would appear on hover, and a single click on any section would return to "normal" mode. This is likely one of those things we'll need to "feel" to make a decision on.

Ok, I think I'll always prefer "click" over "hover" though :P

**Comment by youknowriad** [comment]
Made a lot of updates here, let me know what you think.

**Comment by jameskoster** [comment]
Here's what I'm seeing:


https://user-images.githubusercontent.com/846565/163383360-aeacc94b-e8b6-4931-b5c4-ff867959076e.mp4



* It's strange that list view stays open because you cannot select child blocks.
* If we're only going to reveal toolbars on click then we still need a hover effect to make it obvious the sections are interactive.
* Noticing an occasional on-canvas flicker moving between selections.
* Header toolbar is missing movers for some reason 🤔 
* I really miss the animation on block re-ordering (I know we won't get to that here, just a note).

Edit: The video above didn't demonstrate the flicker, see below (around 5s):

https://user-images.githubusercontent.com/846565/163383814-4022d02f-daed-4b15-b1ef-68056d0a7030.mp4

It's tricky to replicate, I'm not sure what is the trigger.

**Comment by youknowriad** [comment]
> Header toolbar is missing movers for some reason

Is this because it's locked?

> I really miss the animation on block re-ordering (I know we won't get to that here, just a note).

Felt the same, that will be my first follow-up. It's not complex, I just want to make it right (move stuff around) without making this PR huge.

> It's strange that list view stays open because you cannot select child blocks.

I guess, I can just close it for now. We can also decide to leave it functional later. (You select a child, its root parent gets highlighted)

> If we're only going to reveal toolbars on click then we still need a hover effect to make it obvious the sections are interactive.

Makes sense.

> Noticing an occasional on-canvas flicker moving between selections.

The flickering is actually a nightmare in Safari. But these bugs are the hardest to track and fix 😬

**Comment by jameskoster** [comment]
> Is this because it's locked?

Derp, yes. Will just need to style the toolbar so that it doesn't look strange. Or perhaps include the lock icon?

Edit: On second thoughts, definitely include the lock icon. It would be frustrating to have to exit this mode in order to unlock a block you want to change.

> I guess, I can just close it for now.

I think that would be best. There may be a more unique visual treatment we can apply in this mode.

**Comment by youknowriad** [comment]
I think I addressed all the points including the flickering.

**Comment by jameskoster** [comment]
Nice! 😁

I made a couple of small style adjustments. Some other thoughts:

* The focus styles on the movers and the lock are awkwardly aligned, but this seems to be an issue on trunk too, albeit less apparent
* Do we need the Inserter above the first block? It pushes everything down a bit.
* Should the inserters be blue to match other inline inserters?
* Should we disable the Inspector as we've disabled List View? I don't think it serves much purpose in this mode.

**Comment by youknowriad** [comment]
> Do we need the Inserter above the first block? It pushes everything down a bit.

I can remove it but we do need the space in the top to show the first block's toolbar, so maybe we should just keep it?

**Comment by youknowriad** [comment]
> Should the inserters be blue to match other inline inserters?

Some of our inserters are blue others are black. I never know :)

**Comment by youknowriad** [comment]
> Should we disable the Inspector as we've disabled List View? I don't think it serves much purpose in this mode.

This is a bit tricky. I'm not sure I'll be able to do that simply. I can close the sidebar easily on open but disabling entirely is not easy (cc @jorgefilipecosta as it's a complementary area)

**Comment by jorgefilipecosta** [comment]
> > Should we disable the Inspector as we've disabled List View? I don't think it serves much purpose in this mode.
> 
> This is a bit tricky. I'm not sure I'll be able to do that simply. I can close the sidebar easily on open but disabling entirely is not easy (cc @jorgefilipecosta as it's a complementary area)

I guess we can not include the slot on the exploded mode?

**Comment by youknowriad** [comment]
> I guess we can not include the slot on the exploded mode?

I wanted to disable the button, not remove it entirely (like the other header buttons)

**Comment by jorgefilipecosta** [comment]
> > I guess we can not include the slot on the exploded mode?
> 
> I wanted to disable the button, not remove it entirely (like the other header buttons)

This is a more challenging issue. One solution would be to add an optional prop to ComplementaryArea called pinnedUI and when passed that property would be the toggle button of the sidebar, allowing one to have full control and custom logic for when that button on the sidebar is enable or disabled or not rendered or rendered in a different way. We can also go for a simple property if we don't want to allow that level of control like isTooglingDisabled and when true the button would be disabled.

**Comment by youknowriad** [comment]
@jameskoster maybe it's fine for now to just leave it there, or alternatively remove the button entirely?

**Comment by jameskoster** [comment]
Yeah probably fine to leave it for now. At least most of the block settings are hidden.

**Comment by jameskoster** [comment]
> Some of our inserters are blue others are black. I never know :)

I think these ones should be blue :D

**Comment by jasmussen** [comment]
This is cool! And it _sounds_ like we can add animation in the future, that's great!

> I think these ones should be blue :D

I'd echo that. But — hear me out — _what if they were all blue?_ (⊃｡•́‿•̀｡)⊃━☆ﾟ.*･｡ﾟ

Can we have an inserter after the last block as well?

Screenshot 2022-04-15 at 08 14 23

**Comment by youknowriad** [comment]
> This is cool! And it sounds like we can add animation in the future, that's great!

We can add an animation when moving blocks in the view but between the "exploded" and "normal" view, I'm doubtful (or if you're optimistic, I'm unsure at the moment)

**Comment by youknowriad** [comment]
I added an inserter at the end and made all inserters show patterns by default.

File: packages/edit-site/src/components/block-list-exploded/index.js [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+import { store as blockEditorStore, Inserter } from '@wordpress/block-editor';
 import { useSelect } from '@wordpress/data';
+import { pure } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
 import BlockListExplodedItem from './item';
 
+const InserterBeforeBlock = pure( ( { clientId } ) => (
```
Comment (Mamaduka) [inline packages/edit-site/src/components/block-list-exploded/index.js:13]: I didn't know we had this HoC. Should I prefer using this over `React.memo`?

Comment (youknowriad) [reply packages/edit-site/src/components/block-list-exploded/index.js:13]: To be honest at this point, I don't know. What's certain is that we need to make a decision :P. This HoC was here before React introduced React.memo and if I'm not wrong also works for Class component contrary to React.memo which I think is only suited for function components.

File: packages/edit-site/src/components/block-list-exploded/item.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			) }
+			<div
+				ref={ blockWrapper }
+				role="button"
+				onClick={ ( event ) => {
+					if ( event.detail === 1 ) {
+						selectBlock( clientId );
+					} else if ( event.detail === 2 ) {
+						switchEditorMode( 'visual' );
+					}
```
Comment (Mamaduka) [inline packages/edit-site/src/components/block-list-exploded/item.js]: I got a little confused here 😄  I assume `1` and `2` mean single and double click. Maybe we should use constants here.

Comment (youknowriad) [reply packages/edit-site/src/components/block-list-exploded/item.js]: I can use booleans to make things clearer :)

**Comment by jasmussen** [comment]
I know this sounds like a small thing just focusing on the animation, but I want to poke at it a little more to make sure we're on the right path. 

Conceptually I don't think of this view as still being the editor view, but in a different mode. Just like [Browse Mode](#36667) where blocks become inert and navigable, these are still the same blocks, just inert and draggable. In that light, I don't know if replacing all the blocks with previews is the right approach. I know the following inspector hack is not representative of the impressive technical challenge, but here the transform is applied to the styles wrapper, and a few background colors are shuffled around:
Screenshot 2022-04-15 at 09 51 00

This is the "before" state, for comparison:

before

That is to say, I think the animation is important, but more important than that, is the flowing in and out of this state. It isn't conceptualized as being a _shift_ — it's meant to be much more similar to select mode. Smooth flowing in and out.

**Comment by Mamaduka** [comment]
Sometimes when I toggle "Explode" mode, not all sections are rendered. It usually happens after the third or fourth toggle without page refresh.

I also noticed that the editor remembers that it's in "Explode" mode. Is this intentional?


CleanShot 2022-04-15 at 11 43 52

File: packages/edit-site/src/components/block-list-exploded/item.js (hunk 2 above)
Comment (Mamaduka) [reply packages/edit-site/src/components/block-list-exploded/item.js]: Thanks, Riad.

File: packages/edit-site/src/components/block-list-exploded/index.js (hunk 1 above)
Comment (Mamaduka) [reply packages/edit-site/src/components/block-list-exploded/index.js:13]: Right, React.memo works only with functional components.

> What's certain is that we need to make a decision

Got it. Use `React.memo` with function components use `pure` with Class components 😄

**Comment by youknowriad** [comment]
> Conceptually I don't think of this view as still being the editor view, but in a different mode. Just like [Browse Mode](https://github.com/WordPress/gutenberg/pull/40314#36667) where blocks become inert and navigable, these are still the same blocks, just inert and draggable. In that light, I don't know if replacing all the blocks with previews is the right approach. I know the following inspector hack is not representative of the impressive technical challenge, but here the transform is applied to the styles wrapper, and a few background colors are shuffled around:

I'm curious to know what changes you did to the inspector. I agree with you that there were two paths:

 - One is using the initial block list and "style it" to look different.
 - Two is to use a separate view and animate between both.

It's also true that for the first one it's easier to animate things but I don't think it's the right approach for the two things highlighted in the descriptions and the main one for me is this one:

Here we want to apply the editor styles (iframe) to each block separately but we don't want the editor styles to bleed out of the block. And our goal with iframing everything is just to use the "styles" from the theme as is. Meaning if the block itself is not an iframe, we can't apply these styles.

I think trying to use the existing BlockList to show this mode is going to be a hack that will always be very fragile.

**Comment by youknowriad** [comment]
There's also one question here: Do we want to "scale down" the size of the preview according to a given viewportWidth (current state of the PR) or do we want to always render it as is (no scaling down). I tried both, I preferred the scaled version but I can also do "unscaled".

**Comment by youknowriad** [comment]
> Sometimes when I toggle "Explode" mode, not all sections are rendered. It usually happens after the third or fourth toggle without page refresh.

I've seen this happen. I think it's a bug in the `BlockPreview` component, we also see it sometimes in the patterns inserter (height: 0) but it's very hard to reproduce and debug. Could use help there if you manage to do it. 

> I also noticed that the editor remembers that it's in "Explode" mode. Is this intentional?

Initially unintended but I thought I might just well keep it (like the code editor switch)

**Comment by youknowriad** [comment]
@jasmussen I'm exploring the alternative approach here https://github.com/WordPress/gutenberg/pull/40376 I'm not sure yet how I feel about it. I don't have the animation there yet but it's easier to implement for sure.

**Comment by youknowriad** [comment]
closing in favor of https://github.com/WordPress/gutenberg/pull/41156

**Pull request closed without merging**